### PR TITLE
Sanity Portable Text & Site Schemas

### PR DESCRIPTION
**Pull request by Aerilym**

### CODE REVIEW

File: packages/sanity-cms/components/SanityImage.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  isInline,
+  client,
+}: {
+  value: SanityImageSource & { alt?: string };
+  isInline: boolean;
+  client: SessionSanityClient;
+}) => {
+  const { width, height } = getImageDimensions(value);
+  return (
+    <img
```
Comment (yougotwill) [inline packages/sanity-cms/components/SanityImage.tsx:16]: I assume you want to keep a regular img and dont want to use next/image?

File: packages/sanity-cms/components/SanityButton.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  if (!href || !label) {
+    logger.warn('SanityButton: Missing href or label');
+    return null;
+  }
+
+  return (
+    <NavLink href={href}>
+      <Button data-testid={`button:sanity-${label}`} variant="secondary" rounded="md">
```
Comment (yougotwill) [inline packages/sanity-cms/components/SanityButton.tsx:38]: Do we not want to support custom data-testids?

File: packages/sanity-cms/api/validate-link.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  const validateLinkHandler = async (req: NextRequest) => {
+    const urlToCheck = req.nextUrl.searchParams.get('urlToCheck') as string;
+
+    if (!urlToCheck) {
+      return new NextResponse('Missing urlToCheck', { status: 400 });
+    }
+
+    logger.info(`Checking URL: ${urlToCheck}`);
+    const [error, checkedRes] = await safeTry(fetch(urlToCheck));
```
Comment (yougotwill) [inline packages/sanity-cms/api/validate-link.ts:22]: Seeing the usefulness of safeTry!

**Review by yougotwill — COMMENTED** [review]
Left so comments. I wanted to have a look at how the studio renders but realised the staking site doesn't use use the sanity-cms package yet.

It might be nice to have a demo page on "/" and access to the studio via "/admin". 

Probably a TODOFUTURE.

File: packages/ui/tailwind.config.ts [hunk 4]
```diff
@@ -27,6 +27,7 @@ export default {
     },
     fontFamily: {
       'roboto-flex': ['var(--font-roboto-flex)'],
+      'source-serif': ['var(--font-source-serif)'],
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/ui/tailwind.config.ts:30]: Used on the foundation site. coming soon

File: packages/ui/fonts/fonts.ts [hunk 5]
```diff
@@ -28,6 +28,12 @@ const RobotoFlex = Roboto_Flex({
   variable: '--font-roboto-flex',
 });
 
+const SourceSerif = Source_Serif_4({
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/ui/fonts/fonts.ts:31]: Used on the foundation site. coming soon

File: packages/sanity-cms/schemas/social.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    validation: (Rule) => Rule.required(),
+  }),
+  ...linkFields,
+];
+
+export const socialSchema = defineType({
+  type: 'document',
+  name: 'social',
+  title: 'Social',
+  icon: TwitterIcon,
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/sanity-cms/schemas/social.ts:44]: just shows a twitter icon on the CMS, not used by the schema

File: packages/sanity-cms/schemas/site.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    title: 'Show Social Links in Footer',
+    description: 'Enable to show social links in the footer',
+  }),
+  defineField({
+    title: 'Footer Social Links',
+    name: 'footerSocialLinks',
+    description: 'Links to CMS content or external resources',
+    group: 'footer',
+    type: 'array' as const,
+    of: [externalLinkFieldDefinition, internalLinkFieldDefinition, socialLinkFieldDefinition],
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/sanity-cms/schemas/site.ts:54]: supports picking of the 3, very cool :)

File: packages/sanity-cms/schemas/fields/basic/url.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { defineField } from 'sanity';
+
+const checkedUrlResults = new Map<string, string | true>();
+
+export const urlField = defineField({
+  name: 'url',
+  type: 'url',
+  title: 'URL',
+  description:
+    'An external link. If you want to link to an internal link (page or blog post), you must use a reference.',
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/sanity-cms/schemas/fields/basic/url.ts:10]: like reference the page, will probably need to come up with better error messages and descriptions for marketing

File: packages/sanity-cms/queries/getSiteSettings.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+  if (err) {
+    logger.error(err);
+    return null;
+  }
+
+  const siteSettings = result[0];
+
+  if (!siteSettings) {
+    logger.info(`Site settings not found`);
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/sanity-cms/queries/getSiteSettings.ts:25]: site url?

File: packages/sanity-cms/components/SanityImage.tsx (hunk 1 above)
Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx:16]: yeah unfortunate reality of this :(

File: packages/sanity-cms/components/SanityButton.tsx (hunk 2 above)
Comment (Aerilym) [reply packages/sanity-cms/components/SanityButton.tsx:38]: not from sanity, dont want people picking the datatest ids, the label is probably the most stable field for cms buttons

File: packages/sanity-cms/queries/getSiteSettings.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (yougotwill) [reply packages/sanity-cms/queries/getSiteSettings.ts:25]: In the case where the site settings get accidently deleted on the studio or we are starting a new project and we check the logs on say a vps would be good to know which project it is if we have many.

For example `Site settings for token.getsession.org are not found`

File: packages/sanity-cms/schemas/fields/basic/url.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (yougotwill) [reply packages/sanity-cms/schemas/fields/basic/url.ts:10]: Cool, add a TODO comment if you want

File: packages/sanity-cms/components/SanityImage.tsx (hunk 1 above)
Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx:16]: actually im doing some testing with next/image, will update with result

File: packages/sanity-cms/queries/getSiteSettings.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/sanity-cms/queries/getSiteSettings.ts:25]: Ah right, so these queries are used by specific next apps, (eg: static gen of pages) so if this was to log an error you'd know based on the app you were running. We also don't have a way to get the url on the server-side so this is a non-starter.

[the remark this answers is not in the document]
Comment (yougotwill) [reply packages/sanity-cms/queries/getSiteSettings.ts:25]: Cool resolving

File: packages/sanity-cms/components/SanityImage.tsx (hunk 1 above)
Comment (yougotwill) [reply packages/sanity-cms/components/SanityImage.tsx:16]: Cool

File: packages/ui/fonts/fonts.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (yougotwill) [reply packages/ui/fonts/fonts.ts:31]: Cool resolving

File: packages/sanity-cms/components/SanityImage.tsx (hunk 1 above)
Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx]: It now uses next/image

File: packages/sanity-cms/components/SanityImage.tsx [hunk 10]
```diff
@@ -0,0 +1,115 @@
+import urlBuilder from '@sanity/image-url';
+import type { SessionSanityClient } from '../lib/client';
+import Image from 'next/image';
+import logger from '../lib/logger';
+import { getPlaiceholder } from 'plaiceholder';
+import type {
```
Comment (yougotwill) [inline packages/sanity-cms/components/SanityImage.tsx:6]: Typo here. You will need to update the original file as well.
``suggestion
import { getPlaceholder } from 'placeholder';
```

File: packages/sanity-cms/components/SanityImage.tsx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  value,
+  isInline,
+  client,
+  cover,
+  renderWithPriority,
+  className,
+}: SanityImageProps) => {
+  let imageData = {
+    src: '',
+    height: 0,
```
Comment (yougotwill) [inline packages/sanity-cms/components/SanityImage.tsx:50]: Confirming, but you are happy with for whatever reason the image fails to load then a 0x0 image is rendered that we can't see? As opposed to sort some of placeholder div with a background for example?

File: packages/sanity-cms/components/SanityImage.tsx [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+type SanityImageProps = {
+  value: SanityImageType;
+  client: SessionSanityClient;
+  isInline?: boolean;
+  cover?: boolean;
+  renderWithPriority?: boolean;
+  className?: string;
+};
+export const SanityImage = async ({
```
Comment (yougotwill) [inline packages/sanity-cms/components/SanityImage.tsx:40]: Wonder if we should have support for figures with captions?

```tsx
<figure>
  <SanityImage>
  <caption>Description: An image about something really useful</caption>
</figure>
```

File: packages/sanity-cms/package.json [hunk 13]
```diff
@@ -12,7 +12,8 @@
   "scripts": {
     "check-types": "tsc --noEmit",
     "lint": "eslint .",
-    "test": "jest --passWithNoTests"
+    "test": "jest --passWithNoTests",
+    "expose-port": "lt --port 3000"
```
Comment (yougotwill) [inline packages/sanity-cms/package.json:16]: I guess for accessing the studio separately from nextjs on a local machine?

File: packages/sanity-cms/package.json [hunk 14]
```diff
@@ -32,15 +35,20 @@
...
     "@sanity/preview-kit": "^5.1.1",
     "@sanity/preview-url-secret": "^1.6.21",
     "@sanity/vision": "^3.58.0",
     "@session/logger": "workspace:*",
+    "@session/ui": "workspace:*",
     "@session/util-js": "workspace:*",
     "@session/util-logger": "workspace:*",
+    "mini-xml": "^1.2.2",
     "next-sanity": "^9.5.0",
+    "plaiceholder": "^3.0.0",
```
Comment (yougotwill) [inline packages/sanity-cms/package.json:51]: Just a note that this project was archived 31/05/2023. https://github.com/joe-bell/plaiceholder

File: packages/sanity-cms/components/SanityImage.tsx (hunk 10 above)
Comment (yougotwill) [reply packages/sanity-cms/components/SanityImage.tsx:6]: LMFAO ignore this I see this is package 😂 Left my comment for a laugh

File: packages/sanity-cms/package.json (hunk 13 above)
Comment (Aerilym) [reply packages/sanity-cms/package.json:16]: Yeah more specifically to test webhooks

File: packages/sanity-cms/package.json (hunk 14 above)
Comment (Aerilym) [reply packages/sanity-cms/package.json:51]: Yeah I was concerned at first but this is fine, they are the first developer I've ever seen to actually finish a project
> The plaiceholder project is feature complete and will now be kept in maintenance mode.

File: packages/sanity-cms/components/SanityImage.tsx (hunk 10 above)
Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx:6]: lmao yeah this bothered me

File: packages/sanity-cms/components/SanityImage.tsx (hunk 11 above)
Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx:50]: So in any place image are used they don't flow content, except when used in the body. So in the tile component you'll have a blank tile, but in a blog post the image will just be gone

Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx:50]: It think this is better than having a blank image in the middle of a post

Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx:50]: Keeping in mind this only happens if the image fails to build, which is a bad case. This does not happen if an image doesn't exist as the image component is only used if an image is in the cms

File: packages/sanity-cms/components/SanityImage.tsx (hunk 12 above)
Comment (Aerilym) [reply packages/sanity-cms/components/SanityImage.tsx:40]: that would be a separate component, if we want captioned images we can make it easily with your example above.

File: packages/sanity-cms/components/SanityImage.tsx (hunk 11 above)
Comment (yougotwill) [reply packages/sanity-cms/components/SanityImage.tsx:50]: Cool happy with your explanation. Resolving

**Review by yougotwill — APPROVED** [review]
LGTM 🚢

**Pull request merged**